Mathlib/Algebra/Lie/Basic.lean

```diff
@@ -0,0 +1,1134 @@
+
+See note [reducible non-instances]. -/
+@[reducible]
+def LieRingModule.compLieHom : LieRingModule L₁ M
+    where
+  bracket x m := ⁅f x, m⁆
+  lie_add x := lie_add (f x)
+  add_lie x y m := by simp only [LieHom.map_add, add_lie]
+  leibniz_lie x y m := by simp only [lie_lie, sub_add_cancel, LieHom.map_lie]
+#align lie_ring_module.comp_lie_hom LieRingModule.compLieHom
+
+theorem LieRingModule.compLieHom_apply (x : L₁) (m : M) :
+    haveI := LieRingModule.compLieHom M f
+    ⁅x, m⁆ = ⁅f x, m⁆ :=
+  rfl
+#align lie_ring_module.comp_lie_hom_apply LieRingModule.compLieHom_apply
+
+/-- A Lie module may be pulled back along a morphism of Lie algebras.
+
+See note [reducible non-instances]. -/
+@[reducible]
+def LieModule.compLieHom [Module R M] [LieModule R L₂ M] :
+    @LieModule R L₁ M _ _ _ _ _ (LieRingModule.compLieHom M f) :=
+  { LieRingModule.compLieHom M f with
+    smul_lie := fun t x m => by
+      simp only [LieRingModule.compLieHom_apply, smul_lie, LieHom.map_smul]
+    lie_smul := fun t x m => by
+      simp only [LieRingModule.compLieHom_apply, lie_smul] }
+#align lie_module.comp_lie_hom LieModule.compLieHom
+
+end ModulePullBack
+
+/-- An equivalence of Lie algebras is a morphism which is also a linear equivalence. We could
+instead define an equivalence to be a morphism which is also a (plain) equivalence. However it is
+more convenient to define via linear equivalence to get `.toLinearEquiv` for free. -/
+structure LieEquiv (R : Type u) (L : Type v) (L' : Type w) [CommRing R] [LieRing L] [LieAlgebra R L]
+  [LieRing L'] [LieAlgebra R L'] extends L →ₗ⁅R⁆ L' where
+  /-- The inverse function of an equivalence of Lie algebras -/
+  invFun : L' → L
+  /-- The inverse function of an equivalence of Lie algebras is a left inverse of the underlying
+  function. -/
+  left_inv : Function.LeftInverse invFun toLieHom.toFun
+  /-- The inverse function of an equivalence of Lie algebras is a right inverse of the underlying
+  function. -/
+  right_inv : Function.RightInverse invFun toLieHom.toFun
+#align lie_equiv LieEquiv
+
+@[inherit_doc]
+notation:50 L " ≃ₗ⁅" R "⁆ " L' => LieEquiv R L L'
+
+namespace LieEquiv
+
+variable {R : Type u} {L₁ : Type v} {L₂ : Type w} {L₃ : Type w₁}
+
+variable [CommRing R] [LieRing L₁] [LieRing L₂] [LieRing L₃]
+
+variable [LieAlgebra R L₁] [LieAlgebra R L₂] [LieAlgebra R L₃]
+
+/-- Consider an equivalence of Lie algebras as a linear equivalence. -/
+def toLinearEquiv (f : L₁ ≃ₗ⁅R⁆ L₂) : L₁ ≃ₗ[R] L₂ :=
+  { f.toLieHom, f with }
+#align lie_equiv.to_linear_equiv LieEquiv.toLinearEquiv
+
+instance hasCoeToLieHom : Coe (L₁ ≃ₗ⁅R⁆ L₂) (L₁ →ₗ⁅R⁆ L₂) :=
+  ⟨toLieHom⟩
+#align lie_equiv.has_coe_to_lie_hom LieEquiv.hasCoeToLieHom
+
+instance hasCoeToLinearEquiv : Coe (L₁ ≃ₗ⁅R⁆ L₂) (L₁ ≃ₗ[R] L₂) :=
+  ⟨toLinearEquiv⟩
+#align lie_equiv.has_coe_to_linear_equiv LieEquiv.hasCoeToLinearEquiv
+
+/-- Coercion to a function. -/
+@[coe] def toFun' (e : L₁ ≃ₗ⁅R⁆ L₂) := e.toFun
```

Comment:
I have introduced these `toFun'` definitions (following a pattern elsewhere) to create a target for the `@[coe]` decorator so that the `CoeFun` expressions are not eagerly unfolded (which, amongst other things, interferes with simp-normal form).

An alternative would be to use `FunLike` and `EquivLike`. In fact @LukasMias did this for the `FunLike` cases but I reverted his changes in https://github.com/leanprover-community/mathlib4/pull/2262/commits/f4dca383e9126b3901c6b6ac627a88d95bcdedc3 because I expect future porting work will be easier without this change.